mars/services/task/execution/ray/executor.py

```diff
@@ -527,3 +525,71 @@ def _get_subtask_output_keys(chunk_graph: ChunkGraph):
             else:
                 output_keys[chunk.key] = 1
         return output_keys.keys()
+
+    async def _get_newly_finished_subtasks(
+        self,
+        unfinished_subtasks: Set[Subtask],
+        subtask_to_first_output_object_ref: Dict[Subtask, "ray.ObjectRef"],
+    ):
+        unready_object_refs = [
+            subtask_to_first_output_object_ref[subtask]
+            for subtask in unfinished_subtasks
+        ]
+        total = len(unready_object_refs)
+        ready_objects = []
+        if total > 0:
+            ready_objects, _ = await asyncio.to_thread(
+                ray.wait,
+                unready_object_refs,
+                num_returns=total,
+                timeout=0,
+                fetch_local=False,
+            )
+        return set(
+            [
+                self._cur_stage_first_output_object_ref_to_subtask[object_ref]
+                for object_ref in ready_objects
+            ]
+        )
+
+    async def _check_subtask_results_periodically(
+        self,
+        subtask_graph: SubtaskGraph,
+        submitted_subtasks: List[Subtask],
+        subtask_to_first_output_object_ref: Dict[Subtask, "ray.ObjectRef"],
+        time_interval: float,
+    ):
+        total = len(submitted_subtasks)
+        if total == 0:
+            return
+        unfinished_subtasks = set(submitted_subtasks)
+        while True:
+            if self._cancelled:
+                return
+
+            finished_subtasks = await self._get_newly_finished_subtasks(
+                unfinished_subtasks, subtask_to_first_output_object_ref
+            )
+            unfinished_subtasks -= finished_subtasks
+            unfinished_number = len(unfinished_subtasks)
+            stage_progress = (
+                (total - unfinished_number) / total * self._cur_stage_tile_progress
+            )
+            self._cur_stage_progress = self._pre_all_stages_progress + stage_progress
+
+            # Why does here use finished subtasks instead of submitted subtasks?
+            # If a submitted subtask fails, and we remove the chunks it depends,
+            # then we could not rerun it. We can guarantee the chunks will not
+            # be removed until a subtask runs to completion.
+            for subtask in finished_subtasks:
+                for predecessor in subtask_graph.iter_predecessors(subtask):
+                    if all(
+                        successor in submitted_subtasks
+                        for successor in subtask_graph.iter_successors(predecessor)
+                    ):
+                        for chunk in predecessor.chunk_graph.results:
+                            self._task_context.pop(chunk.key, None)
+
+            if unfinished_number == 0:
```

Comment:
Can we make it simpler?
- [Producer] An aiotask that updates the progress and generates finish subtasks.
- [Consumer] An aiotask that takes finish subtasks. For each finish subtask, check all the successors of the predecessor subtasks are finish.

GC the output object refs of the subtask which successors are submitted (not finish as above) can reduce the memory peaks, but we can't cancel and rerun slow subtasks because the input object refs of running subtasks may be deleted.